types/enzyme/index.d.ts

```diff
@@ -23,7 +23,7 @@ export class ElementClass extends Component<any, any> {
  * all specified in the implementation. TS chooses the EnzymePropSelector overload and loses the generics
  */
 export interface ComponentClass<Props> {
-    new(props?: Props, context?: any): Component<Props, any>;
+    new(props: Props, context?: any): Component<Props>;
```

Comment:
It's possible to write React components with constructors that don't accept props, and the signatures wouldn't match here. Seems this would break overload selection for .find and friends on such components.

Reply:
Hi @jwbay thanks for the review.

Honestly, I'm not sure why this works (which is not a good sign), but it does. My original approach unioned a no-arg, 1-arg, and 2-arg component class (to handle exactly what you point out), but I dropped that approach when I realized this simpler approach actually does work.

E.g. if I change `MyComponent` and `AnotherComponent` to have 0-arg constructors:

```
class MyComponent extends Component<MyComponentProps, MyComponentState> {
    constructor() {
        super({stringProp: "1", numberProp: 2});
    }
    setState(...args: any[]) {
        console.log(args);
    }
}

class AnotherComponent extends Component<AnotherComponentProps> {
    constructor() {
        super({});
    }
    setState(...args: any[]) {
        console.log(args);
    }
}
```

All of the enyzme-tests code still type-checks, including this line that is in this PR:

```
        const s1: EnzymeComponentClass<AnotherComponentProps> = AnotherComponent;
```

Oh...actually, I bet I know why it works; in TS it's legal to implement a method but not declare all of the parameters, e.g.:

```
type Callback = (status: number, done: boolean) => void;
function doSomething(c: Callback) {}
doSomething(() => console.log("done"));
```

The 0-arg closure still matches the 2-arg `Callback` type, by conveniently ignoring the two params.

So, in the same way, a 0-arg `MyComponent.new()` will (somewhat surprisingly but actually conveniently) match the `ComponentClass.new(props, context)` and so the `find<P2>(component: ComponentClass<P2>): ShallowWrapper<P2, any>` overload still kicks in.